### DISCUSSION TITLE
Strange boolean evaluation rules

### DISCUSSION THREAD

**Discussion by travisdowns** [community/community#25645] [Actions]
<p>I’m trying to understand boolean expressions in GitHub actions. The <a href="https://docs.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#literals" rel="noopener nofollow ugc">documentation gives the following example</a> of literals of different types, including boolean as <code>${{ false }}</code> and <code>${{ true }}</code>.</p>
<p>Following their lead I use the <a href="https://github.com/travisdowns/gha-test/blob/8231125ef930b66aaec38561addabcd245a1a036/.github/workflows/build.yml#L25-L35" rel="noopener nofollow ugc">following step</a> with an <code>env</code> section that sets <code>VARF</code> to false and <code>VART</code> to true and then in the <code>run</code> key has some test output:</p>
<pre><code>  - name: Var test
    env:
      VARF: ${{ false }}
      VART: ${{ true }}
    run: |
      echo "VART=${{ env.VARF }}"
      echo "VART=${{ env.VART }}"
      echo "VARF &amp;&amp; VART=${{ env.VARF &amp;&amp; env.VART }}"
      echo "VARF || VART=${{ env.VARF || env.VART }}"
      echo "!VARF=${{ ! env.VARF }}"
      echo "!!VART=${{ !! env.VART }}"
</code></pre>
<p>The <a href="https://github.com/travisdowns/gha-test/runs/3425012281?check_suite_focus=true#step:3:1" rel="noopener nofollow ugc">output</a> is as follows:</p>
<pre><code class="lang-auto">VART=false
VART=true
VARF &amp;&amp; VART=true
VARF || VART=false
!VARF=false
!!VART=true
</code></pre>
<p>The plain output is as expected, showing <code>true</code> and <code>false</code>. The output of the <code>&amp;&amp;</code> and <code>||</code> operators is the opposite of expected, giving <code>false &amp;&amp; true == true</code> and <code>false || true == false</code>. The output of <code>${{ ! env.VARF }}</code> is false when I’d expect true.</p>
<p>What’s going on here? Possibly variables in the env context get coerced to strings, but that still doesn’t explain all the results.</p>

**Comment by jsoref** [comment, marked as answer]
<p><code>env</code> defines strings.</p>
<p>Consider this (node):</p>
<pre><code class="lang-auto">% node
Welcome to Node.js v16.7.0.
Type ".help" for more information.
&gt; "false" &amp;&amp; "true"
'true'
&gt; "false" || "true"
'false'
&gt; !"false"
false
&gt; !"true"
false
&gt; !!"true"
true
</code></pre>
<p>To do what you’re trying to do, you’d probably want something like:</p>
<pre><code class="lang-yaml">      echo "VARF &amp;&amp; VART=${{ fromJSON(env.VARF) &amp;&amp; fromJSON(env.VART) }}"
</code></pre>

**Comment by jsoref** [comment]
<p>Technically <code>env</code> is defined here:<br>
<a href="https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#env" class="onebox" target="_blank" rel="noopener">https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#env</a><br>
And use is covered here:<br>
<a href="https://docs.github.com/en/actions/reference/environment-variables" class="onebox" target="_blank" rel="noopener">https://docs.github.com/en/actions/reference/environment-variables</a></p>
<p>Practically, while it says they’re environment variables, it doesn’t actually say they’re strings, it just kinda assumes you’d figure that out.</p>
<p>If it were my project, I wouldn’t object to someone filing a bug/PR asking for the term <code>string</code> to be sprinkled around the various bits.</p>
<p>(It isn’t my project.)</p>
